src/Microsoft.AspNetCore.Routing/Dispatcher/RouterDispatcher.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Dispatcher;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing.Dispatcher
+{
+    /// <summary>
+    /// An adapter to plug an <see cref="IRouter"/> into a dispatcher.
+    /// </summary>
+    public class RouterDispatcher : DispatcherBase
+    {
+        private readonly IRouter _router;
+
+        public RouterDispatcher(IRouter router)
+        {
+            if (router == null)
+            {
+                throw new ArgumentNullException(nameof(router));
+            }
+
+            _router = router;
+        }
+
+        public async override Task InvokeAsync(HttpContext httpContext)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            var routeContext = new RouteContext(httpContext);
+            await _router.RouteAsync(routeContext);
```

Comment:
This isn't really **all** we need to do to plug this in. We have to full-on fake the `RouteData` including a fake route collection. We also have to create the router feature.

I'm feeling pretty good about this though. I think it won't be super complicated to plug in routing in a super compatible way. Then anyone using `IRouter` based exensibility doesn't have to throw it away.